File: 1283-Find the Smallest Divisor Given a Threshold.py
class Solution(object):
    def smallestDivisor(self, nums, threshold):
        """
        :type nums: List[int]
        :type threshold: int
        :rtype: int
        """
        right = max(nums)
        left = 1
        
        def getPieces(mid):
            count = 0
            for i in range(len(nums)):
                count += -(-nums[i]//mid)  # how did someone come up with this??
            return count
        
        while left + 1 < right:
            mid = left + (right-left)//2
            if getPieces(mid) > threshold:
                left = mid
            else:
                right = mid
 
        if getPieces(left) <= threshold:
            return left
        else:
            return right

File: JavaScript/1283-Find the Smallest Divisor Given a Threshold.js
/**
 * @param {number[]} nums
 * @param {number} threshold
 * @return {number}
 */
var smallestDivisor = function(nums, threshold) {
    let left = 1, right = Math.max(...nums);
    
    let getCount = (mid) =>{
        let count = 0;
        for (let num of nums){
            count += Math.ceil(num/mid);
        }
        return count;
    }
    
    while (left + 1 < right){
        let mid = left + Math.floor((right-left)/2)
        if (getCount(mid) > threshold){
            left = mid;
        } else{
            right = mid;
        }
    }
    if (getCount(left) <= threshold){
        return left;
    }else{
        return right;
    }
};